core/src/raw/oio/read/buffer_reader.rs

```diff
@@ -0,0 +1,457 @@
+                return Poll::Ready(Err(err));
+            }
+            return match ready!(self.r.poll_read(cx, dst)) {
+                Ok(n) => {
+                    self.cur += n as u64;
+                    Poll::Ready(Ok(n))
+                }
+                Err(err) => Poll::Ready(Err(err)),
+            };
+        }
+
+        let (read_start, read_end) = (self.cur as usize, self.cur as usize + dst.len());
+
+        if let Some((buf_start, buf_end)) = self.buffer_range() {
+            if buf_start == buf_end && buf_start == self.cur as usize {
+                self.clear_buffer();
+                return Poll::Ready(Ok(0));
+            }
+            // Hit the buffer.
+            //     +---------------+
+            //     | Buffer Data   |
+            // ----+^--------------^-----
+            //      |              |
+            //      |              |
+            //      cur            cur + dst.len()
+            if buf_start <= read_start && read_end <= buf_end {
+                // Invariant: buf.len() <= cap
+                let offset = self.cur as usize - buf_start;
+                let amt = min(buf_end - self.cur as usize, dst.len());
+
+                dst.put_slice(&self.buf[offset..offset + amt]);
+                self.cur = read_end as u64;
+                return Poll::Ready(Ok(amt));
+            } else if buf_start <= read_start && read_start < buf_end {
+                // Hit part of buffer.
+                //     +---------------+
+                //     | Buffer Data   |
+                // ----+--------^----------^--
+                //              |          |
+                //              |          |
+                //              cur        cur + dst.len()
+
+                let offset = self.cur as usize - buf_start;
+                let rem = min(buf_end - self.cur as usize, dst.len());
+
+                dst.put_slice(&self.buf[offset..offset + rem]);
+
+                if let Err(err) = ready!(self.poll_seek(cx, SeekFrom::Start(buf_end as u64))) {
+                    return Poll::Ready(Err(err));
+                }
+                if self.has_reach_end() {
+                    return Poll::Ready(Ok(rem));
+                }
+            } else if let Err(err) = ready!(self.poll_seek(cx, SeekFrom::Start(self.cur))) {
+                return Poll::Ready(Err(err));
+            }
+        }
+
+        // Continue read
+        match ready!(self.poll_fill_buf(cx)) {
+            Ok(_) => self.poll_read(cx, dst),
+            Err(err) => Poll::Ready(Err(err)),
+        }
+    }
+
+    fn poll_seek(&mut self, cx: &mut Context<'_>, pos: SeekFrom) -> Poll<Result<u64>> {
+        match ready!(self.r.poll_seek(cx, pos)) {
+            Ok(cur) => {
+                self.cur = cur;
+
+                Poll::Ready(Ok(cur))
+            }
+            Err(err) => Poll::Ready(Err(err)),
+        }
+    }
+
+    fn poll_next(&mut self, cx: &mut Context<'_>) -> Poll<Option<Result<Bytes>>> {
```

Comment:
poll_next should respect buffer.

Reply:
> poll_next should respect buffer.

Sorry, I didn't get this. Forwarding the buffer directly seems sensible.

Reply:
> Sorry, I didn't get this. Forwarding the buffer directly seems sensible.

For example, users called `read` and `next`. It's possible that we have buffered some data internally, we should return the buffered data first.